### PR TITLE
Support install addons in customized namespaces

### PR DESCRIPTION
**Pull request by xuezhaojun**
Signed-off-by: xuezhaojun <zxue@redhat.com>

Epic: 
https://github.com/stolostron/backlog/issues/23093

User-story:
https://github.com/stolostron/backlog/issues/23929

### CODE REVIEW

File: pkg/spoke/addon/namespace_controller.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	"k8s.io/apimachinery/pkg/labels"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	addoninformerv1alpha1 "open-cluster-management.io/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "open-cluster-management.io/api/client/addon/listers/addon/v1alpha1"
+)
+
+const (
+	addonInstallNamespace = "addon.open-cluster-management.io/namespace"
```
Comment (qiujian16) [inline pkg/spoke/addon/namespace_controller.go]: addonInstallNamespaceAnnotationKey

File: pkg/spoke/addon/namespace_controller.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+				_, err = c.kubeClient.CoreV1().Namespaces().Update(ctx, ns, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// When addon is deleted, check if there is anyother addon is using the same installNamespace, if not, delete the namespace
+	if !addOn.DeletionTimestamp.IsZero() {
```
Comment (qiujian16) [inline pkg/spoke/addon/namespace_controller.go:78]: I am worried about deleting ns here.

File: pkg/spoke/addon/namespace_controller.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+) factory.Controller {
+	c := &addonNamespaceController{
+		kubeClient:  kubeClient,
+		addOnLister: addOnInformer.Lister(),
+		recorder:    recorder,
+	}
+	return factory.New().WithInformersQueueKeyFunc(func(o runtime.Object) string {
+		accessor, _ := meta.Accessor(o)
+		return accessor.GetName()
+	}, addOnInformer.Informer()).WithSync(c.sync).ToController("AddonNamespaceController", recorder)
```
Comment (qiujian16) [inline pkg/spoke/addon/namespace_controller.go:46]: should also watch ns also right? Also you would need registration to have permission to create ns. And update permission in registration-operator at first.

File: pkg/spoke/addon/namespace_controller.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	if err != nil {
+		return err
+	}
+
+	// Get installNamespace of managedClusterAddon
+	installNamespace := addOn.Spec.InstallNamespace
+
+	// Check installNamespace exist or not
+	ns, err := c.kubeClient.CoreV1().Namespaces().Get(ctx, installNamespace, metav1.GetOptions{})
```
Comment (qiujian16) [inline pkg/spoke/addon/namespace_controller.go]: use resourceApply.ApplyNamespace

**Review by qiujian16 — COMMENTED** [review]
I do not think we should always delete the ns when addon is missing. Maybe we should have another annotation on ns so ns will not be deleted.

File: pkg/spoke/addon/namespace_controller.go (hunk 1 above)
Comment (xuezhaojun) [reply pkg/spoke/addon/namespace_controller.go]: Fixed.

File: pkg/spoke/addon/namespace_controller.go (hunk 3 above)
Comment (xuezhaojun) [reply pkg/spoke/addon/namespace_controller.go:46]: the permission is updated in this PR https://github.com/open-cluster-management-io/registration-operator/pull/253 . But not sure about watch ns.

File: pkg/spoke/addon/namespace_controller.go (hunk 4 above)
Comment (xuezhaojun) [reply pkg/spoke/addon/namespace_controller.go]: Changed to use resourceapply pkg.

File: pkg/spoke/addon/namespace_controller.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+				addonInstallNamespaceAnnotationKey: "true",
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// When addon is deleted, check if there is anyother addon is using the same installNamespace, if not, delete the namespace
+	if !addOn.DeletionTimestamp.IsZero() {
```
Comment (skeeey) [inline pkg/spoke/addon/namespace_controller.go:78]: we may put this at the beginning

File: pkg/spoke/addon/namespace_controller.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		addOnList, err := c.addOnLister.ManagedClusterAddOns(c.managedClusterName).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		for _, a := range addOnList {
```
Comment (skeeey) [inline pkg/spoke/addon/namespace_controller.go:83]: exclude current addon?

**Comment by qiujian16** [comment]
/hold 
since we had some further discussion. This may not be needed.

**Comment by xuezhaojun** [comment]
/close
Change to the approach that addons are responsible for creating ns they need.

**Pull request closed without merging**